TCP2: Add support for arbitrary data payloads

ozhuraki opened the pull request:
At present, TCP2 can’t correctly handle payloads greater than 128 bytes.

So, add support for arbitrary data payloads.

As a result, eliminate net_pkt_linearize().

ozhuraki committed:
net: tcp2: Fix the clang compilation warnigs for native_posix

Fix the clang compilation warnigs for the native_posix target:

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Fix the snprintf() format specification

Zephyr's snprintf() doesn't support %zd.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: buf: Use __ASSERT() in NET_BUF_ASSERT()

NET_BUF_ASSERT() tranlates into nothing in presence
of CONFIG_ASSERT=y, which is misleading.

Use __ASSERT() in NET_BUF_ASSERT().

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Add a support for net_buf reservation

This simplifies the error handling in buffer allocation.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Support arbitrary payloads in TCP windows

Support arbitrary payloads in TCP windows.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Support tracking of net_buf_clone()

Support tracking of net_buf_clone().

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Support arbitrary payloads in tcp_csum().

Support arbitrary payloads in tcp_csum().

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Delete unused tcp_pkt_linearize()

Delete unused tcp_pkt_linearize().

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

tbursztyka commented on subsys/net/ip/nbr.c:
```diff
@@ -179,7 +179,7 @@ struct net_nbr *net_nbr_lookup(struct net_nbr_table *table,
 {
-	NET_ASSERT_INFO(idx < CONFIG_NET_IPV6_MAX_NEIGHBORS,
+	NET_ASSERT(idx < CONFIG_NET_IPV6_MAX_NEIGHBORS,
 			"idx %d >= max %d", idx,
```
you'll need to realign all the parameters to last '(' + 1 space

tbursztyka commented on samples/net/sockets/echo/src/socket_echo.c:
```diff
@@ -73,7 +73,7 @@ int main(void)
 
 		while (1) {
-			char buf[128], *p;
+			static char buf[8192], *p;
```
can you tell why in the commit message? (besides reducing recv() calls)
Because it may affect some targets with limited amount of ram.

tbursztyka commented on subsys/net/buf.c:
```diff
@@ -35,7 +35,7 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define NET_BUF_WARN(fmt, ...)
 #define NET_BUF_INFO(fmt, ...)
-#define NET_BUF_ASSERT(cond)
+#define NET_BUF_ASSERT(cond, ...) __ASSERT(cond, "" __VA_ARGS__)
```
it's changing the scope of this macro. Can you tell in your commit message why you want that?

tbursztyka commented on subsys/net/buf.c:
```diff
@@ -340,6 +341,7 @@ struct net_buf *net_buf_alloc_len(struct net_buf_pool *pool, size_t size,
 	pool->avail_count--;
 	NET_BUF_ASSERT(pool->avail_count >= 0);
 #endif
+	NET_BUF_ASSERT(req_size <= size);
```
insert this assert after line 329

tbursztyka commented on subsys/net/buf.c:
```diff
@@ -235,6 +235,7 @@ struct net_buf *net_buf_alloc_len(struct net_buf_pool *pool, size_t size,
 #endif
 {
 	u32_t alloc_start = k_uptime_get_32();
+	size_t req_size = size;
```
move this to line 316

tbursztyka commented on subsys/net/ip/tcp2.c:
```diff
@@ -34,6 +34,49 @@ static void tcp_in(struct tcp *conn, struct net_pkt *pkt);
+
+	while (rsv_bytes < len) {
+
+		buf = tcp_nbuf_alloc(conn, 128);
```
why 128? and could you use a macro for it instead (with a comment summarizing its usage)

tbursztyka reviewed (commented):
Your commit message would greatly gain with more details added. When it's trivial (like copyright update) I agree, but many others are really specific and it would accelerate the review to get more details there.

ozhuraki committed:
net: buf: Add an assert in net_buf_alloc_len()

Add an assert in net_buf_alloc_len() for undersized alloc.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Use CONFIG_NET_BUF_DATA_SIZE for network buffer size

Use CONFIG_NET_BUF_DATA_SIZE for the maximum network buffer size
in TCP windows.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki commented:
@tbursztyka

Thanks for the input, fixed.

Regarding 506ec7a, if there’s a reason to translate NET_BUF_ASSERT() into nothing, I can drop it.

What’s the logic, anyway?

tbursztyka commented:
@ozhuraki indeed assert is not a log level. Instead, put it out of the ifdef, and make it use __ASSERT as you did. If one does not want such assert to be made, it will be possible to disable it via unselecting CONFIG_ASSERT 
Or create a NET_ASSERT in Kconfig... either way. Let's see what @jhedberg thinks about it

ozhuraki commented:
@tbursztyka

OK and done.

jhedberg commented:
> Let's see what @jhedberg thinks about it

The mapping to __ASSERT seems fine to me

jukkar commented on subsys/net/ip/tcp2.c:
```diff
@@ -192,7 +192,8 @@ static const char *tcp_th(struct net_pkt *pkt)
 
 	if (th->th_off < 5) {
-		s += snprintf(s, buf_size, "Bogus th_off: %hu", th->th_off);
+		s += snprintf(s, buf_size, "Bogus th_off: %hu",
```
I wonder if we should use `snprintk()` as that is used more in networking code. Only caveat is that the snprintk() does not support all the modifiers.

jukkar commented on subsys/net/ip/tcp2.c:
```diff
@@ -487,47 +487,57 @@ static const char *tcp_state_to_str(enum tcp_state state, bool prefix)
 
-	NET_ASSERT(len, "Zero length data");
+	while (total) {
+		len = (total <= 128) ? total : 128;
```
Where is the 128 value coming, we should use define instead.

jukkar commented on subsys/net/ip/tp.c:
```diff
@@ -165,6 +165,24 @@ struct net_buf *tp_nbuf_alloc(struct net_buf_pool *pool, size_t len,
+				const char *func)
+{
+	struct net_buf *clone = net_buf_clone(buf, K_NO_WAIT);
+	struct tp_nbuf *tb = k_malloc(sizeof(struct tp_nbuf));
```
Probably unlikely to happen but we should check the return value of k_malloc()

jukkar commented on subsys/net/ip/tcp2.c:
```diff
@@ -722,17 +722,29 @@ static uint16_t cs(int32_t s)
 static void tcp_csum(struct net_pkt *pkt)
```
Can't we use the generic checksum calculation functions that exist already?

jukkar commented on subsys/net/ip/tcp2.c:
```diff
@@ -18,6 +18,8 @@ LOG_MODULE_REGISTER(net_tcp, CONFIG_NET_TCP_LOG_LEVEL);
 #include "net_private.h"
 #include "tcp2_priv.h"
 
+#define TCP_NET_BUF_SIZE 128
```
Shouldn't we use CONFIG_NET_BUF_DATA_SIZE here instead of hard coded 128?

ozhuraki committed:
net: tcp2: Use CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT

Use CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT for the
retransmission timeout value.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Drop snprintf() and use snprintk()

Drop snprintf() and use snprintk() as snprintk()
is more suitable for the networking code.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

jukkar reviewed (approved):
Generally LGTM, although usually we try "fix" the original commits and not to introduce the fixes in new commits in the PR.

jhedberg commented:
What's the status of this PR? @ozhuraki are you going to address the latest comment from @jukkar? Fixes to commits that are not yet upstream should be done in those commits themselves (using e.g. `git rebase -i`) rather than by adding new "fixup" commits to the PR.

@tbursztyka are you ok with this PR after the latest changes?

ozhuraki commented:
@jhedberg

> What's the status of this PR?

Feedback had been addressed. The PR itself might have been missed while the master have been blocked during the release.

> Fixes to commits that are not yet upstream should be done in those commits themselves (using e.g. git rebase -i) rather than by adding new "fixup" commits to the PR.

Fixes had been done to the commits themselves. That comment might be slightly out of date, unless I am missing something.

jhedberg commented:
@ozhuraki ok, thanks! @tbursztyka are you still planning to review this?

tbursztyka reviewed (dismissed):
Please apply the comment I made (you still have parameters to re-indent in first or second patch) and reply to our questions about hard coded numbers etc...

ozhuraki committed:
net: core: Drop NET_ASSERT_INFO() macro

A single assert macro can serve a purpose here.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki commented:
@tbursztyka

> Please apply the comment I made (you still have parameters to re-indent in first or second patch)

Sorry, a few have slipped indeed, corrected.

> and reply to our questions about hard coded numbers etc...

The rest had been fixed.

tbursztyka commented:
ah, instead of fixing in place the hard coded number you added another patch to do it everywher. ok.

jukkar commented:
Ok, lets get this merged as there are more TCP patches coming after this one.
